### PR TITLE
Move to lowdb & use across entire HF

### PR DESCRIPTION
**Pull request by f3rno**
Description pending; see new files for the structure. Model example:

```js
const MODEL_TYPES = require('../model_types')
const FIELD_TYPES = require('../field_types')
const genModel = require('./util/gen_model')

module.exports = genModel({
  name: 'Algorithmic Order',
  path: 'algo_orders',
  type: MODEL_TYPES.MAP,
  schema: {
    gid: FIELD_TYPES.STRING,
    algoID: FIELD_TYPES.STRING,
    state: FIELD_TYPES.OBJECT,
    active: FIELD_TYPES.BOOLEAN
  },

  key: ({ algoID, gid } = {}) => `${algoID}:${gid}`,

  extraMethods: ({ getAll, update }) => ({
    getActive: () => getAll().filter({ active: true }).value(),
    idString: (ao) => `${ao.algoID}:${ao.gid}`,

    start: (ao) => (
      update(ao, v => ({
        ...v,
        active: true
      }))
    ),

    stop: (ao) => (
      update(ao, v => ({
        ...v,
        active: false
      }))
    ),
  })
})
```

Model & document validation is handled by 2 helper methods. Example output from a test script:
Screenshot from 2019-05-20 16-18-43

The idea is to use a lightweight DB and code helper methods in-house. MongoDB/NEMS has issues in production, sqlite is too limited (column editing is unsupported in migrations, etc).

### CODE REVIEW

**Comment by f3rno** [comment]
Output of `node examples/create.js` w/ full verbose logging:

```
➜  bfx-hf-models git:(refactor-switch-db-lowdb) node examples/create.js 
  bfx:hf:models:validate:model (Algorithmic Order) has path (algo_orders) [OK] +0ms
  bfx:hf:models:validate:model (Algorithmic Order) path length is 11 [OK] +2ms
  bfx:hf:models:validate:model (Algorithmic Order) schema is object [OK] +0ms
  bfx:hf:models:validate:model (Algorithmic Order) checking for schema field validator gid (isString) [OK] +0ms
  bfx:hf:models:validate:model (Algorithmic Order) checking for schema field validator algoID (isString) [OK] +0ms
  bfx:hf:models:validate:model (Algorithmic Order) checking for schema field validator state (isObject) [OK] +0ms
  bfx:hf:models:validate:model (Algorithmic Order) checking for schema field validator active (isBoolean) [OK] +0ms
  bfx:hf:models:validate:model (HF Backtest) has path (backtests) [OK] +0ms
  bfx:hf:models:validate:model (HF Backtest) path length is 9 [OK] +0ms
  bfx:hf:models:validate:model (HF Backtest) schema is object [OK] +0ms
  bfx:hf:models:validate:model (HF Backtest) checking for schema field validator indicators (isObject) [OK] +0ms
  bfx:hf:models:validate:model (HF Backtest) checking for schema field validator trades (isArray) [OK] +0ms
  bfx:hf:models:validate:model (HF Backtest) checking for schema field validator symbol (isString) [OK] +0ms
  bfx:hf:models:validate:model (HF Backtest) checking for schema field validator tf (isString) [OK] +0ms
  bfx:hf:models:validate:model (HF Backtest) checking for schema field validator from (_isDateHF) [OK] +1ms
  bfx:hf:models:validate:model (HF Backtest) checking for schema field validator to (_isDateHF) [OK] +0ms
  bfx:hf:models:validate:model (HF Backtest) checking for schema field validator btID (isString) [OK] +0ms
  bfx:hf:models:validate:model (HF Backtest) checking for schema field validator strategyID (isString) [OK] +0ms
  bfx:hf:models:validate:model (Trade) has path (trades) [OK] +0ms
  bfx:hf:models:validate:model (Trade) path length is 6 [OK] +0ms
  bfx:hf:models:validate:model (Trade) schema is object [OK] +0ms
  bfx:hf:models:validate:model (Trade) checking for schema field validator bfxID (isFinite) [OK] +0ms
  bfx:hf:models:validate:model (Trade) checking for schema field validator price (isFinite) [OK] +0ms
  bfx:hf:models:validate:model (Trade) checking for schema field validator amount (isFinite) [OK] +0ms
  bfx:hf:models:validate:model (Trade) checking for schema field validator mts (isFinite) [OK] +0ms
  bfx:hf:models:validate:model (Trade) checking for schema field validator symbol (isString) [OK] +0ms
  bfx:hf:models:validate:model (Candle) has path (candles) [OK] +0ms
  bfx:hf:models:validate:model (Candle) path length is 7 [OK] +0ms
  bfx:hf:models:validate:model (Candle) schema is object [OK] +0ms
  bfx:hf:models:validate:model (Candle) checking for schema field validator open (isFinite) [OK] +0ms
  bfx:hf:models:validate:model (Candle) checking for schema field validator high (isFinite) [OK] +0ms
  bfx:hf:models:validate:model (Candle) checking for schema field validator low (isFinite) [OK] +0ms
  bfx:hf:models:validate:model (Candle) checking for schema field validator close (isFinite) [OK] +0ms
  bfx:hf:models:validate:model (Candle) checking for schema field validator vol (isFinite) [OK] +0ms
  bfx:hf:models:validate:model (Candle) checking for schema field validator symbol (isString) [OK] +0ms
  bfx:hf:models:validate:model (Candle) checking for schema field validator tf (isString) [OK] +0ms
  bfx:hf:models:validate:model (Candle) checking for schema field validator type (isString) [OK] +0ms
  bfx:hf:models:validate:model (Candle) checking for schema field validator mts (isFinite) [OK] +0ms
  bfx:hf:models:db:init:db loading from /home/f3rno/code/bfx/hf/bfx-hf-models/lib/init/../../db/test.json +28ms
  bfx:hf:models:examples:create creating AO w/ ID test-algo-id, GID test-gid +10ms
  bfx:hf:models:util:gen-map-model [Algorithmic Order] set on key algo_orders.test-algo-id-test-gid +2ms
  bfx:hf:models:examples:create fetching AO... +6ms
  bfx:hf:models:util:gen-map-model [Algorithmic Order] get key algo_orders.test-algo-id-test-gid +0ms
  bfx:hf:models:examples:create read AO {
  "gid": "test-gid",
  "algoID": "test-algo-id",
  "active": false,
  "state": {
    "someAlgoState": 42,
    "moreState": {
      "nestedFields": {
        "mts": "2019-05-21T09:04:35.855Z"
      }
    }
  }
} +1ms
  bfx:hf:models:examples:create cleaning up... +0ms
  bfx:hf:models:util:gen-map-model [Algorithmic Order] remove key algo_orders.test-algo-id-test-gid +0ms
  bfx:hf:models:examples:create AO deleted. confirming... +0ms
  bfx:hf:models:util:gen-map-model [Algorithmic Order] get key algo_orders.test-algo-id-test-gid +0ms
  bfx:hf:models:examples:create read AO: undefined +0ms
```

**Comment by f3rno** [comment]
Model definitions look like this:
```js
const MODEL_TYPES = require('../model_types')
const FIELD_TYPES = require('../field_types')
const genModel = require('../util/gen_model')

module.exports = genModel({
  name: 'Algorithmic Order',
  path: 'algo_orders',
  type: MODEL_TYPES.MAP,
  schema: {
    gid: FIELD_TYPES.STRING,
    algoID: FIELD_TYPES.STRING,
    state: FIELD_TYPES.OBJECT,
    active: FIELD_TYPES.BOOLEAN
  },

  key: ({ algoID, gid } = {}) => `${algoID}-${gid}`,
  extraMethods: ({ getAll }) => ({
    getActive: () => getAll().filter({ active: true }).value()
  })
})
```

File: lib/models/candle/sync_range.js [hunk 1]
```diff
@@ -1,18 +1,89 @@
...
+    // sync entire range
+    if (existingCandles.length === 0) {
+      boundsToSync.push({ start, end })
+    } else { // sync to fill gaps
+      const lastCandle = _last(existingCandles)
+      const gapData = auditGaps({ type, symbol, tf }, { start, end })
+      const { gaps } = gapData
+
+      // missing after known candles
+      if (lastCandle.mts !== end) {
```
Comment (JacobPlaster) [inline lib/models/candle/sync_range.js]: `lastCandle.mts < end` is probably better than `lastCandle.mts !== end`

Comment (JacobPlaster) [reply lib/models/candle/sync_range.js]: Actually I think this should be remove completely and left to the `auditGaps` function to determine?

File: lib/models/candle/audit_gaps.js [hunk 2]
```diff
@@ -0,0 +1,26 @@
+const { TIME_FRAME_WIDTHS } = require('bfx-hf-util')
+const _isFinite = require('lodash/isFinite')
+
+module.exports = ({ getInRange }) => (doc, { start, end }) => {
+  const gaps = []
+  const { tf } = doc
+  const candles = getInRange(doc, { start, end })
+
+  if (candles.length < 2) {
+    return { gaps, candles }
```
Comment (JacobPlaster) [inline lib/models/candle/audit_gaps.js:10]: Seems stratnge to return empty gaps here? If the candles returned are less than 2 maybe we should just consider the whole thing a gap?

i.e:

```
if (candles.length < 2) {
    gaps.push({start, end})
    return { gaps, candles: [] }
}
```

File: lib/models/candle/sync_range.js [hunk 3]
```diff
@@ -1,18 +1,89 @@
...
+const debug = require('debug')(DEBUG_STR)
+const { RESTv2 } = require('bfx-api-node-rest')
+const PromiseThrottle = require('promise-throttle')
+const PI = require('p-iteration')
+const _last = require('lodash/last')
+const Promise = require('bluebird')
+
+const FETCH_LIMIT = 1000
+const rest = new RESTv2({ transform: true })
+const pt = new PromiseThrottle({
```
Comment (JacobPlaster) [inline lib/models/candle/sync_range.js:12]: Nice, I wonder if we should add an exponential backoff retry just incase we do get rate limited?

File: lib/models/candle/audit_gaps.js [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+  const width = TIME_FRAME_WIDTHS[tf]
+
+  if (!_isFinite(width)) {
+    throw new Error(`invalid time frame [unknown width]: ${tf}`)
+  }
+
+  for (let i = 0; i < candles.length - 1; i += 1) {
+    if ((candles[i].mts - candles[i + 1].mts) !== width) {
+      gaps.push(i)
```
Comment (JacobPlaster) [inline lib/models/candle/audit_gaps.js:21]: We could improve the efficiency of this later on by stitching together gaps if there are less than X candles inbetween. For example we can get 4k candles in a single request so if we can cover multiple gaps in a single 4k request then we could just forget about the candles we have inbetween.

Comment (f3rno) [reply lib/models/candle/audit_gaps.js:21]: Good point, but keep in mind this is meant for internal generic use to ensure there is a candle for each point in time in a range. The req limit is external to this logic; we can always add a `width` arg in the future to override the internal width.

File: lib/models/candle/audit_gaps.js (hunk 2 above)
Comment (f3rno) [reply lib/models/candle/audit_gaps.js:10]: I was initially going to include this, but due to my comment on your other remark I'm keeping it out; if there are no candles, there are no gaps by candle width (this method is meant to check for consistency, not presence.)

File: package.json [hunk 5]
```diff
@@ -22,13 +25,17 @@
...
   "dependencies": {
+    "bfx-api-node-rest": "^1.0.9",
+    "bfx-api-node-util": "^1.0.0",
     "bfx-hf-util": "^1.0.0",
+    "binary-search-insert": "^1.0.3",
     "bluebird": "^3.5.1",
     "copy": "^0.3.1",
     "lodash": "^4.17.10",
+    "lowdb": "^1.0.0",
     "mongoose": "^5.2.16",
```
Comment (JacobPlaster) [inline package.json]: remove mongoose?

File: package.json [hunk 6]
```diff
@@ -22,13 +25,17 @@
...
     "nems": "f3rno/nodejs-embedded-mongodb-standalone",
```
Comment (JacobPlaster) [inline package.json]: remove `nems`?

File: test/lib/models/method_generators/rm_all.js [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+    const currentModels = getAll()
+    assert(currentModels.length === 0) // rmAll check
+
+    // restore
+    initialModels.forEach(m => insert(m))
+
+    const finalModels = getAll()
+    assert.deepStrictEqual(finalModels, initialModels)
+  })
+})
```
Comment (JacobPlaster) [inline test/lib/models/method_generators/rm_all.js]: not a big change, but make sure you enable the white space at end of file feature

**Review by JacobPlaster — APPROVED** [review]

File: test/lib/models/method_generators/rm_all.js (hunk 7 above)
Comment (f3rno) [reply test/lib/models/method_generators/rm_all.js]: Done, I ran `standard --fix`

**Pull request merged**